### PR TITLE
(PDB-1450) Improved node expiration

### PR DESCRIPTION
**Pull request by mullr**
This includes several commits related to improving our handling of deactivated nodes. It's based on PDB-1310, so until that gets merged, a per-commit review may be in order.

### CODE REVIEW

File: src/puppetlabs/puppetdb/command.clj [hunk 1]
```diff
@@ -213,13 +214,18 @@
...
+  (let [certname (:certname payload)
+        producer_timestamp (to-timestamp (:producer_timestamp payload (now)))
+        id  (:id annotations)
+        newer-record-exists? (fn [entity] (scf-storage/have-record-produced-after? entity certname producer_timestamp))]
+    (jdbc/with-transacted-connection db
+      (when-not (scf-storage/certname-exists? certname)
+        (scf-storage/add-certname! certname))
+      (when (not-any? newer-record-exists? [:catalogs :factsets :reports])
+        (scf-storage/deactivate-node! certname producer_timestamp)))
+    (log/info (format "[%s] [%s] %s" id (command-names :deactivate-node) certname))))
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/command.clj]: I think `log/info (format` is equivalent to `log/infof`

File: test/puppetlabs/puppetdb/scf/storage_test.clj [hunk 2]
```diff
@@ -1226,7 +1226,21 @@
...
+    (is (= (map :certname (query-to-vec "SELECT certname FROM certnames ORDER BY certname ASC"))
+           ["node1" "node3"]))))
+
+(deftest purge-expired-nodes
+  (testing "should purge nodes which were expired before the specified date"
+    (add-certname! "node1")
+    (add-certname! "node2")
+    (add-certname! "node3")
+    (expire-node! "node1")
+    (with-redefs [now (constantly (ago (days 10)))]
```
Comment (ajroetker) [inline test/puppetlabs/puppetdb/scf/storage_test.clj]: nitpick, I usually think the clj-time stuff is more readable with `(-> 10 days ago)`

**Comment by mullr** [comment]
@pljenkinsro retest this please

File: documentation/api/query/v4/nodes.markdown [hunk 3]
```diff
@@ -15,8 +15,8 @@ Nodes can be queried by making an HTTP request to the `/nodes` endpoint.
 
 ## `GET /v4/nodes`
 
-This will return all nodes matching the given query. Deactivated nodes
-aren't included in the response.
+This will return all nodes matching the given query. Deactivated and expired
+nodes aren't included in the response.
```
Comment (wkalt) [inline documentation/api/query/v4/nodes.markdown]: we need to find a good place to define the term "expired"

File: src/puppetlabs/puppetdb/scf/storage.clj [hunk 4]
```diff
@@ -1186,6 +1196,16 @@
...
+(pls/defn-validated expire-node!
+  "Expire the given host, recording the current time. If the node is
+  currently expired, no change is made."
+  [certname :- String & [timestamp :- pls/Timestamp]]
+  (let [timestamp (to-timestamp (or timestamp (now)))]
+   (sql/do-prepared "UPDATE certnames SET expired = ?
+                    WHERE certname=? AND expired IS NULL"
+                    [timestamp certname])))
+
+
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/scf/storage.clj]: extra line here

**Comment by kbarber** [comment]
@mullr this needs a rebase BTW.

File: test/puppetlabs/puppetdb/scf/storage_test.clj (hunk 2 above)
Comment (mullr) [inline test/puppetlabs/puppetdb/scf/storage_test.clj]: I agree... but the whole file is like `(ago (days 10))`. I'll make a (maint) commit to clean this up.

File: documentation/api/query/v4/nodes.markdown (hunk 3 above)
Comment (mullr) [inline documentation/api/query/v4/nodes.markdown]: Yes, we do. Maybe in to config docs, when we talk about node-ttl?

File: src/puppetlabs/puppetdb/http/query.clj [hunk 5]
```diff
@@ -11,25 +11,48 @@
...
 
-(defn add-criteria [crit query]
+(defn query-criteria
+  "Extract the 'criteria' (select) part of the given query"
+  [query]
   (cm/match [query]
-            [["extract" columns (expr :guard nil?)]]
-            ["extract" columns crit]
+    [["extract" _ expr]] expr
+    :else (or query nil)))
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/http/query.clj]: `(or query nil)` is that somehow not equivalent to `query`?

**Comment by mullr** [comment]
@pljenkinsro retest this please

File: documentation/api/query/v4/nodes.markdown (hunk 3 above)
Comment (wkalt) [inline documentation/api/query/v4/nodes.markdown]: yeah, maybe. might be worth here as well since that goes into the mechanics a bit https://docs.puppetlabs.com/puppetdb/latest/maintain_and_tune.html

I feel like we additionally should have some kind of glossary page for this kind of stuff long term. Out of scope for this ticket though.

File: documentation/configure.markdown [hunk 6]
```diff
@@ -280,9 +280,11 @@ This controls how often, in minutes, to compact the database. The compaction pro
 
 ### `node-ttl`
 
-Auto-deactivate nodes that haven't seen any activity (no new catalogs,
-facts, etc) in the specified amount of time. You may specify the time
-as a string using any of the following suffixes:
+Mark as 'expired' nodes that haven't seen any activity (no new catalogs,
+facts, etc) in the specified amount of time. Expired nodes behave similarly
+to deactivated nodes; they are excluded from query results by default.
```
Comment (wkalt) [inline documentation/configure.markdown]: deactivated nodes aren't excluded on all endpoints (events for instance) so we might not want to make that claim

Comment (mullr) [inline documentation/configure.markdown]: Ergh, right you are.

File: src/puppetlabs/puppetdb/http/factsets.clj [hunk 7]
```diff
@@ -26,7 +26,7 @@
                http-q/restrict-query-to-active-nodes)}
 
    [node "facts" &]
-   (comp (facts/facts-app version) (partial http-q/restrict-query-to-node node))))
+   (comp (facts/facts-app version false) (partial http-q/restrict-query-to-node node))))
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/http/factsets.clj]: is this a change of behavior?

Comment (mullr) [inline src/puppetlabs/puppetdb/http/factsets.clj]: Yes; you can now get a node's facts even if the node is deactivated or expired.

File: src/puppetlabs/puppetdb/http/query.clj [hunk 8]
```diff
@@ -11,25 +11,48 @@
...
+
+(defn is-active-node-criteria? [criteria]
+  (cm/match [criteria]
+    [["=" ["node" "active"] _]] criteria
+    :else false))
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/http/query.clj]: we also support queries like ["null?", "expired", true/false] . Should we be picking those up as well here?

Comment (wkalt) [inline src/puppetlabs/puppetdb/http/query.clj]: I should say, we say we support them. They currently don't work on the deactivated field due to restrictions.

Comment (mullr) [inline src/puppetlabs/puppetdb/http/query.clj]: I'm not sure if we should or not. My impression from the query engine was
that `[= [node active] true]` was special, so I wouldn't have to worry
about the general case.

File: src/puppetlabs/puppetdb/query/population.clj [hunk 9]
```diff
@@ -12,7 +12,7 @@
   {:post [(number? %)]}
   (-> (str "SELECT COUNT(*) AS c "
            "FROM catalogs clogs, catalog_resources cr, certnames c "
-           "WHERE clogs.id=cr.catalog_id AND c.certname=clogs.certname AND c.deactivated IS NULL")
+           "WHERE clogs.id=cr.catalog_id AND c.certname=clogs.certname AND c.deactivated IS NULL AND c.expired IS NULL")
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/query/population.clj]: I think you should be able to just break this string over multiple lines (without str or multiple quotes)

File: src/puppetlabs/puppetdb/query/population.clj [hunk 10]
```diff
@@ -38,7 +38,7 @@
   {:post [(number? %)]}
   (let [num-unique (-> (query-to-vec (str "SELECT COUNT(*) AS c FROM "
                                           "(SELECT DISTINCT resource FROM catalog_resources cr, catalogs clogs, certnames c "
-                                          " WHERE cr.catalog_id=clogs.id AND clogs.certname=c.certname AND c.deactivated IS NULL) r"))
+                                          " WHERE cr.catalog_id=clogs.id AND clogs.certname=c.certname AND c.deactivated IS NULL AND c.expired IS NULL) r"))
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/query/population.clj]: likewise here

File: documentation/api/query/v4/nodes.markdown (hunk 3 above)
Comment (wkalt) [inline documentation/api/query/v4/nodes.markdown]: we should probably make a note of that in the upgrading guide.

Comment (wkalt) [inline documentation/api/query/v4/nodes.markdown]: sorry, that comment ^ is on the wrong thread

File: src/puppetlabs/puppetdb/http/factsets.clj (hunk 7 above)
Comment (wkalt) [inline src/puppetlabs/puppetdb/http/factsets.clj]: we should probably make a note of that in the upgrading guide.

**Comment by wkalt** [comment]
+1 aside from some minor comments

File: src/puppetlabs/puppetdb/query/population.clj (hunk 9 above)
Comment (mullr) [inline src/puppetlabs/puppetdb/query/population.clj]: but that doesn't indent *nearly* as well.

Comment (mullr) [inline src/puppetlabs/puppetdb/query/population.clj]: I'll clean up this file in a follow-on maint commit

File: src/puppetlabs/puppetdb/http/factsets.clj (hunk 7 above)
Comment (mullr) [inline src/puppetlabs/puppetdb/http/factsets.clj]: Updated the upgrading guide and the endpoint docs.

File: src/puppetlabs/puppetdb/http/query.clj (hunk 8 above)
Comment (mullr) [inline src/puppetlabs/puppetdb/http/query.clj]: I'm going with "leave it as-is" here to avoid scope creep. We can fix that query separately if we want to.

**Comment by mullr** [comment]
@pljenkinsro retest this please
